document/core/syntax/types.rst

```diff
@@ -73,39 +73,84 @@ Conventions
 
 
 
-.. index:: ! heap type, store, type identifier, ! substitution, ! closed type
+.. index:: ! heap type, store, type identifier, ! substitution, ! closed type, ! abstract type, ! concrete type, ! unboxed scalar
    pair: abstract syntax; heap type
+.. _type-i31:
 .. _type-subst:
 .. _type-closed:
+.. _type-abstract:
+.. _type-concrete:
 .. _syntax-heaptype:
 
 Heap Types
 ~~~~~~~~~~
 
 *Heap types* classify objects in the runtime :ref:`store <store>`.
+There are three disjoint hierarchies of heap types:
+
+- *function types* classify :ref:`functions <syntax-func>`,
+- *aggregate types* classify dynamically allocated *managed* data, such as *structures*, *arrays*, or *unboxed scalars*,
+- *external types* classify *external* references possibly owned by the :ref:`embedder <embedder>`.
+
+The values from the latter two hierarchies are interconvertible by ways of the |EXTERNINTERNALIZE| and |EXTERNINTERNALIZE| instructions.
+That is, both type hierarchies are inhabited by an isomorphic set of values, but may have different, incompatible representations in practice.
 
 .. math::
    \begin{array}{llll}
    \production{heap type} & \heaptype &::=&
-     \FUNC ~|~ \EXTERN ~|~ \typeidx ~|~ \functype ~|~ \BOT \\
+     \FUNC ~|~ \NOFUNC \\&&|&
+     \EXTERN ~|~ \NOEXTERN \\&&|&
+     \ANY ~|~ \EQ ~|~ \I31 ~|~ \STRUCT ~|~ \ARRAY ~|~ \NONE \\&&|&
+     \typeidx ~|~ \deftype ~|~ \BOT \\
    \end{array}
 
-The type |FUNC| denotes the infinite union of all types of :ref:`functions <syntax-func>`, regardless of their concrete :ref:`function types <syntax-functype>`.
+A heap type is either *abstract* or *concrete*.
+
+The abstract type |FUNC| denotes the infinite union (i.e., supertype) of all :ref:`function types <syntax-functype>`, regardless of their concrete definition.
+Dually, the type |NOFUNC| denotes the infinite intersection (i.e., subtype) of all :ref:`function types <syntax-functype>`, regardless of their concrete definition.
+This type has no values.
```

Comment:
It seems complicated to describe these in terms of unions and intersections. Would dropping that part and just using supertype and subtype to describe them work?